### ISSUE TITLE
Fix center or zoom

### ISSUE THREAD

**Issue by finaiized** [istarkov/google-map-react#162]
How would I make the center or the zoom level of the map unchangeable by the user? I read through #87 about controlled components and such, so I took this demo (https://jsbin.com/yihohugoza/edit?js,console,output) and removed `setState` under `_onChange`. As I understand, because this is a controlled component, the map should no longer be able to move or zoom as it's center and zoom level are fixed. I tried a similar thing with props, but I can still move around. Am I misunderstanding how this should work?

**Comment by istarkov** [comment]
I wrote somewhere in issues that its really hard to make this control work as any internal React uncontrollable component. As google api is not declarative etc. 
You could prevent zoom change via options prop, there are few ways to do it, for example set minZoom = zoom = maxZoom 
I think there should be option to prevent map moving.

**Comment by finaiized** [comment]
Appreciate the reply. Just to clarify, the component above is a _controlled_ component, right? I'm assuming by specifying a `zoom` and `center` that's like a normal React component's `value`, and because there's an `onChange` handler, it's a controlled component?

The other thing is props. Take this example (https://jsbin.com/xihocihigi/edit?js,console,output) which is from #87 where props are used to determined to zoom and center. You suggested that if `onChange` was added, this becomes a controlled component, but when I try it, it seems to have no effect.

Basically, it seems `zoom` and `center` props aren't really in sync with what you set them to. You can update it but that doesn't always update the map; you don't change them but they update internally.

**Comment by istarkov** [comment]
Below I assume that map always used in controllable manner, so it has `zoom` `center` `onChange` and `onChange` changes the `zoom` and `center` on every call.
In most cases when you change center or zoom properties they should be changed see source https://github.com/istarkov/google-map-react/blob/master/src/google_map.js#L235-L246 but.

But calls to `onChange` callback is created via `idle` google api callback, you could see this here http://istarkov.github.io/google-map-clustering-example/ nothing will occur if you will drag map fast, on only then you stop some markers will occur.

This is because that until `idle` event occurred google api moves layer with React components using his internal api, and this event gives some guarantee that current google map coordinates is the real coordinates. (In other events, map coordinates could be really different than real on the screen, because of internal google api optimizations like `raf` and `animations`)

So coordinates you have in onChange callback is synchronized with real coordinates only on the moment of this callback.

About jsbin I wrote where https://github.com/istarkov/google-map-react/issues/87#issuecomment-171331940 
There is also example with state https://jsbin.com/malovinare/1/edit?js,console,output

So I can't get about what your question is ;-)

**Comment by finaiized** [comment]
Thanks again for the in-depth response.

I think this explains, in a way, why I can't fix the zoom or center in `onChange` or when I pass in zoom/center to the Map as a prop - because `onChange` is only called when `idle` is called. But shouldn't this mean that if I I take the JSBin example you linked and called `setState` like this:

```js
this.setState({
  center: [60.938043, 30.337157],
  zoom: 9
});
```

Shouldn't the map always reset it's zoom and center after I pan and zoom the map and the map comes to idle?

**Comment by istarkov** [comment]
You will get near the same issues as in React if you will set something different on value than onChange has provided.

As an example https://jsbin.com/kajufehezu/edit?js,console,output if you will try to insert character in any place except the end of input, your cursor will always move at the end of the input. 

So here
```javascript
  _onChange = ({center, zoom}) => {
    this.setState({
      center: [61.938043, 31.337157],
      zoom: zoom,      
    });
  }
  ```
 is the same issue - something will go wrong ;-)

**Comment by finaiized** [comment]
I did some more experimenting with similar solutions and came to the same conclusion - things don't always work when you leave the React realm to make changes to DOM, or Maps in this case. Really appreciate your explanations @istarkov, I think I just need to adopt a slightly different mindset with how things work with dealing with Maps. 👌
